addons/l10n_sa_invoice/models/account_move.py

```diff
@@ -0,0 +1,53 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+import base64
+
+from odoo import api, fields, models, _
+from odoo.exceptions import UserError
+
+
+class AccountMove(models.Model):
+    _inherit = 'account.move'
+
+    l10n_sa_delivery_date = fields.Date(string='Delivery Date', default=fields.Date.context_today, copy=False)
+    l10n_sa_show_delivery_date = fields.Boolean(compute='_compute_show_delivery_date')
+    l10n_sa_qr_code_str = fields.Char(string='QR Code', compute='_compute_qr_code_str')
+    l10n_sa_confirmation_datetime = fields.Datetime(string='Confirmation Date', readonly=True, copy=False)
+
+    @api.depends('country_code', 'move_type')
+    def _compute_show_delivery_date(self):
+        for move in self:
+            move.l10n_sa_show_delivery_date = move.country_code == 'SA' and move.move_type in ('out_invoice', 'out_refund', 'out_receipt')
+
+    @api.depends('amount_total', 'amount_untaxed', 'l10n_sa_confirmation_datetime', 'company_id', 'company_id.vat')
+    def _compute_qr_code_str(self):
+        """ Generate the qr code for Saudi e-invoicing. Specs are available at the following link at page 23
+        https://zatca.gov.sa/ar/E-Invoicing/SystemsDevelopers/Documents/20210528_ZATCA_Electronic_Invoice_Security_Features_Implementation_Standards_vShared.pdf
+        """
+        def get_qr_encoding(tag, field):
+            company_name_byte_array = field.encode('UTF-8')
+            company_name_tag_encoding = tag.to_bytes(length=1, byteorder='big')
+            company_name_length_encoding = len(company_name_byte_array).to_bytes(length=1, byteorder='big')
+            return company_name_tag_encoding + company_name_length_encoding + company_name_byte_array
+
+        for record in self:
+            qr_code_str = ''
+            if record.l10n_sa_confirmation_datetime and record.company_id.vat:
+                seller_name_enc = get_qr_encoding(1, record.company_id.display_name)
+                company_vat_enc = get_qr_encoding(2, record.company_id.vat)
+                timestamp_enc = get_qr_encoding(3, record.l10n_sa_confirmation_datetime.isoformat())
+                invoice_total_enc = get_qr_encoding(4, str(record.amount_total))
+                total_vat_enc = get_qr_encoding(5, str(record.currency_id.round(record.amount_total - record.amount_untaxed)))
+
+                str_to_encode = seller_name_enc + company_vat_enc + timestamp_enc + invoice_total_enc + total_vat_enc
+                qr_code_str = base64.b64encode(str_to_encode).decode('UTF-8')
+            record.l10n_sa_qr_code_str = qr_code_str
+
+    def action_post(self):
+        for record in self:
+            if record.country_code == 'SA' and record.l10n_sa_delivery_date < record.invoice_date:
+                raise UserError(_('Delivery Date cannot be before Invoice Date'))
+        self.write({
+            'l10n_sa_confirmation_datetime': fields.Datetime.now()
```

Comment:
the field does not accept timezone aware datetimes

Reply:
@bakboka Yes, but it could suppose that the time you are giving is for the timezone of KSA.

Reply:
Does it matter? now() is UTC because the fields is gonna be stored as UTC. But once the user reads it (on the form or the report) it is gonna transform it into saudi time, making it correct. I think that forcing it to be stored in the database as saudi time will make it erroneous as we are removing twice the timezone difference.

The only issue i can see, is if a user print an invoice while being in a different timezone than saudi which should almost never happen